zookeeper-server/src/main/java/org/apache/zookeeper/server/NIOServerCnxn.java

```diff
@@ -235,10 +237,12 @@ void handleWrite(SelectionKey k) throws IOException, CloseRequestException {
                 if (bb == ServerCnxnFactory.closeConn) {
                     throw new CloseRequestException("close requested");
                 }
+                if (bb == packetSentinel) {
```

Comment:
Seems like you're fixing a bug by introducing `packetSentinel`. Is that correct?

Reply:
Since we separate the response into multiple pieces, length, header, data, we need a way to tell if it's the end of a response or not, and the sentinel is added for this purpose.